proposals/permissions-requestSiteAccess-api.md

```diff
@@ -0,0 +1,135 @@
+# permissions.requestSiteAccess() API
+
+## API Overview
+
+### Background
+
+Extensions can request host permissions via:
+- Manifest declared `host_permissions`, granted at install time by default
+- Manifest declared `optional_host_permissions`, granted at runtime after a user gesture by `<browser>.permissions.request()`
+- Manifest declared `matches` for a `content_script`, granted at install time by default
+
+Users can withhold host permissions, causing the extension to only have access when the extension is [invoked](https://developer.chrome.com/docs/extensions/develop/concepts/activeTab#invoking-activeTabhttps://developer.chrome.com/docs/extensions/develop/concepts/activeTab#invoking-activeTab).
+
+Each browser may decide how it signals the user when an extension is requesting access (e.g through the extensions menu). However, there is no way for an extension to explicitly signal at runtime it’s requesting site access after it was withheld without a user gesture and a heavy-weight permission dialog.
+
+
+### Objective
+
+Allow the extension to show site access requests at runtime without any user gesture in a less-obtrusive way than with permissions.request(). This can be done with a new API that:
+
+- Applies to a specific tab or document id
+- Doesn’t need to be made inside the handler for a user action
+- Shows the request in the UI, handled differently by each browser. See more in [UI Elements and User-Visible Effects](#ui-elements-and-user-visible\-effects) section
+- When accepted, grants always access to the site’s origin
+- Resets the request on cross-origin navigation
+
+#### Use Cases
+
+An extension requested access to a site but the user withheld its access and forgot about it. Extension wants to signal it needs site access to execute. For example, “shopping” extension wants to show site access when user navigates to “amazon.com” and access was withheld.
```

Comment:
If an extension does not have access, how does it know what `tabId` is needed to pass to this API? They can't see the `url` or `title` on the tab object if they don't have access (at least in Safari, where those are empty unless you have host permissions, that might be different in Chrome?).

Reply:
Oh, that's an interesting question. In Chrome the `tabs` permission is sufficient to access the URL. In Safari, I wonder if there are other situations where the extension realises it might be useful on a page - based on a shortcut running perhaps? - and the extension could request access. I know Safari has a lot of behaviour to automatically prompt the user though so perhaps not.

Reply:
Out of interest, does the `tabs` permission do anything in Safari? Despite the name it isn't needed to call (I think all) tabs APIs in Chrome.

Reply:
@oliverdunk Yes, `tabs` does signal to us that the extension implicitly wants access to all websites, so they don't need `*://*/*` in the `permissions` to get our implicit permission request dialogs when doing `tabs` actions like `query()`, `executeScript()`, etc.